src/dm/DispatchManagerActions.cc

```diff
@@ -2256,3 +2455,751 @@ int DispatchManager::live_updateconf(std::unique_ptr<VirtualMachine> vm,
+        const RequestAttributes& ra, string& error_str)
+{
+    auto vm = vmpool->get(vid);
+
+    if ( vm == nullptr )
+    {
+        error_str = "Virtual machine does not exist";
+        return -1;
+    }
+
+    if ((vm->get_state() != VirtualMachine::POWEROFF &&
+         vm->get_state() != VirtualMachine::UNDEPLOYED) ||
+        vm->get_lcm_state() != VirtualMachine::LCM_INIT)
+    {
+        error_str = "VM in wrong state, it has to be in poweroff";
+        return -1;
+    }
+
+    int hid;
+
+    if (vm->hasHistory())
+    {
+        hid = vm->get_hid();
+    }
+    else
+    {
+        error_str = "Could not find host information";
+        return -1;
+    }
+
+    unique_ptr<Host> host;
+
+    if (vm->get_state() ==  VirtualMachine::POWEROFF)
+    {
+        HostShareCapacity sr;
+
+        sr.vmid = vid;
+        sr.pci.push_back(pci);
+
+        host = hpool->get(hid);
+
+        if ( host == nullptr )
+        {
+            error_str = "Could not find host information";
+            return -1;
+        }
+
+        if (!host->add_pci(sr))
+        {
+            error_str = "Cannot assign PCI device in host. Check address and free devices";
+            return -1;
+        }
+    }
+
+    if ( vm->attach_pci(pci, error_str) == -1 )
+    {
+        return -1;
+    }
+
+    if (host)
+    {
+        hpool->update(host.get());
+    }
+
+    close_cp_history(vmpool, vm.get(), VMActions::PCI_ATTACH_ACTION, ra);
+
+    vm->log("DiM", Log::INFO, "PCI device successfully attached.");
+
+    vmpool->update_search(vm.get());
+
+    time_t the_time = time(0);
+
+    vm->set_running_etime(the_time);
+
+    vmpool->update_history(vm.get());
+
+    vmpool->update(vm.get());
+
+    return 0;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+int DispatchManager::detach_pci(int vid, int pci_id, const RequestAttributes& ra,
+        string&  error_str)
+{
+    auto vm = vmpool->get(vid);
+
+    if ( vm == nullptr )
+    {
+        error_str = "Virtual machine does not exist";
+        return -1;
+    }
+
+    if ((vm->get_state() != VirtualMachine::POWEROFF &&
+         vm->get_state() != VirtualMachine::UNDEPLOYED) ||
+        vm->get_lcm_state() != VirtualMachine::LCM_INIT)
+    {
+        error_str = "VM in wrong state, it has to be in poweroff";
+        return -1;
+    }
+
+    int hid = -1;
+
+    if (vm->hasHistory())
+    {
+        hid = vm->get_hid();
+    }
+
+    VectorAttribute * vpci = vm->get_pci(pci_id);
+
+    if (vpci == nullptr)
+    {
+        error_str = "PCI device not found";
+        return -1;
+    }
+
+    if ( vpci->vector_value("TYPE") == "NIC" )
+    {
+        error_str = "Use one.vm.detachnic to detach this PCI device";
+        return -1;
+    }
+
+    HostShareCapacity sr;
+
+    sr.vmid = vid;
+    sr.pci.push_back(vpci);
+
+    if (auto host = hpool->get(hid))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This snippet may contain an accidental use of incorrect operation. Make sure that the operation being performed is in fact what you intended.